### PR TITLE
Major compaction refactoring

### PR DESCRIPTION
**Pull request by wayblink**

### CODE REVIEW

File: internal/datacoord/clustering_compaction_job.go [hunk 1]
```diff
@@ -31,54 +31,86 @@ type ClusteringCompactionJob struct {
...
-	completedPlans  []*datapb.CompactionPlan
-	failedPlans     []*datapb.CompactionPlan
-	timeoutPlans    []*datapb.CompactionPlan
-	analysisTaskID  UniqueID
+	state     compactionTaskState
+	startTime uint64
+	endTime   uint64
+	// should only store partial info in meta
+	compactionPlans      []*datapb.CompactionPlan
+	compactionPlanStates []compactionTaskState
```
Comment (xiaocai2333) [inline internal/datacoord/clustering_compaction_job.go:39]: It is recommended to use state as a field of plan.

File: internal/datacoord/clustering_compaction_job.go [hunk 2]
```diff
@@ -31,54 +31,86 @@ type ClusteringCompactionJob struct {
...
+	analysisTaskID       UniqueID
 }
 
 func convertClusteringCompactionJob(info *datapb.ClusteringCompactionInfo) *ClusteringCompactionJob {
```
Comment (xiaocai2333) [inline internal/datacoord/clustering_compaction_job.go:43]: rename to `convertToClusteringCompactionJob` ?

File: internal/datacoord/compaction_trigger.go [hunk 3]
```diff
@@ -594,6 +594,12 @@ func (t *compactionTrigger) handleGlobalSignal(signal *compactionSignal) error {
 }
 
 func (t *compactionTrigger) handleClusteringCompactionSignal(signal *compactionSignal) error {
+	ts, err := t.allocTs()
+	if err != nil {
+		log.Warn("allocate ts failed, skip to handle compaction")
+		return err
+	}
+
 	t.forceMu.Lock()
```
Comment (czs007) [inline internal/datacoord/compaction_trigger.go:603]: can this lock be removed ?

Comment (wayblink) [reply internal/datacoord/compaction_trigger.go:603]: not recommend to remove it. Only one handleXXXCompactionSignal can do at the same time. The following logic are all memory updates.

File: internal/datacoord/clustering_compaction_job.go (hunk 2 above)
Comment (wayblink) [reply internal/datacoord/clustering_compaction_job.go:43]: ok

File: internal/datacoord/meta.go [hunk 4]
```diff
@@ -182,6 +185,14 @@ func (m *meta) reloadFromKV() error {
 		m.clusteringCompactions[fmt.Sprintf("%d-%d", info.CollectionID, info.TriggerID)] = info
 	}
 
+	partitionStatsInfos, err := m.catalog.ListPartitionStatsInfos(m.ctx)
+	if err != nil {
+		return err
+	}
+	for _, info := range partitionStatsInfos {
+		m.partitionStatsInfos[fmt.Sprintf("%d/%d/%s/%d", info.CollectionID, info.PartitionID, info.VChannel, info.Version)] = info
```
Comment (czs007) [inline internal/datacoord/meta.go:193]: why use info.Version and the part of key?

File: internal/datacoord/clustering_compaction_manager.go [hunk 5]
```diff
@@ -114,6 +119,79 @@ func (t *ClusteringCompactionManager) startJobCheckLoop() {
 	}
 }
 
+func (t *ClusteringCompactionManager) startGCLoop() {
+	defer logutil.LogPanic()
+	defer t.wg.Done()
```
Comment (czs007) [inline internal/datacoord/clustering_compaction_manager.go:124]: create gcTicker in here.

File: internal/datacoord/clustering_compaction_manager.go [hunk 6]
```diff
@@ -86,34 +85,115 @@ func (t *ClusteringCompactionManager) stop() {
...
 func (t *ClusteringCompactionManager) startJobCheckLoop() {
 	defer logutil.LogPanic()
 	defer t.wg.Done()
+	ticker := time.NewTicker(paramtable.Get().DataCoordCfg.ClusteringCompactionStateCheckInterval.GetAsDuration(time.Second))
+	defer ticker.Stop()
 	for {
 		select {
 		case <-t.quit:
-			t.ticker.Stop()
+			ticker.Stop()
```
Comment (czs007) [inline internal/datacoord/clustering_compaction_manager.go:108]: unnecessary

File: internal/datacoord/meta.go (hunk 4 above)
Comment (wayblink) [reply internal/datacoord/meta.go:199]: PartitionStats has version. One partition may have multi partitionStats,  clustering_compaction_manager control the  version update and gc logic.

**Pull request merged**